logic/src/commonTest/kotlin/com/wire/kalium/logic/feature/message/ephemeral/EphemeralMessageDeletionHandlerTest.kt

```diff
@@ -0,0 +1,393 @@
+            .with(any(), any(), any())
+            .wasInvoked(exactly = Times(1))
+
+        verify(arrangement.messageRepository)
+            .suspendFunction(arrangement.messageRepository::getMessageById)
+            .with(any(), any())
+            .wasInvoked(exactly = Times(1))
+    }
+
+    @Test
+    fun givenRegularMessage_whenEnqueueingTwice_thenSelfDeletionShouldBeCalledOnce() = runTest(testDispatcher.default) {
+        // given
+        val (arrangement, ephemeralMessageDeletionHandler) = Arrangement(this, testDispatcher)
+            .withMessageRepositoryReturningMessage(
+                message = TestMessage.TEXT_MESSAGE.copy(
+                    expirationData = Message.ExpirationData(
+                        expireAfter = 1.seconds,
+                        selfDeletionStatus = Message.ExpirationData.SelfDeletionStatus.NotStarted
+                    )
+                )
+            )
+            .withMessageRepositoryMarkingSelfDeletionStartDate()
+            .arrange()
+
+        // when
+        ephemeralMessageDeletionHandler.startSelfDeletion(
+            conversationId = ConversationId("someValue", "someDomain"),
+            messageId = "someId"
+        )
+
+        ephemeralMessageDeletionHandler.startSelfDeletion(
+            conversationId = ConversationId("someValue", "someDomain"),
+            messageId = "someId"
+        )
+
+        advanceUntilIdle()
+
+        // then
+        verify(arrangement.messageRepository)
+            .suspendFunction(arrangement.messageRepository::markSelfDeletionStartDate)
+            .with(any(), any(), any())
+            .wasInvoked(exactly = Times(1))
+
+        verify(arrangement.messageRepository)
+            .suspendFunction(arrangement.messageRepository::getMessageById)
+            .with(any(), any())
+            .wasInvoked(exactly = Times(2))
+    }
+
+    @Test
+    fun givenRegularMessageWithExpiration_whenEnqueueForDeletionAndTimeElapsed_thenTheMessageShouldBeDeleted() =
+        runTest(testDispatcher.default) {
+            // given
+            val (arrangement, ephemeralMessageDeletionHandler) = Arrangement(this, testDispatcher)
+                .withMessageRepositoryReturningMessage(
+                    message = TestMessage.TEXT_MESSAGE.copy(
+                        expirationData = Message.ExpirationData(
+                            expireAfter = 1.seconds,
+                            selfDeletionStatus = Message.ExpirationData.SelfDeletionStatus.NotStarted
+                        )
+                    )
+                )
+                .withMessageRepositoryMarkingSelfDeletionStartDate()
+                .withDeletingMessage()
+                .arrange()
+
+            // when
+            ephemeralMessageDeletionHandler.startSelfDeletion(
+                conversationId = ConversationId("someValue", "someDomain"),
+                messageId = "someId"
+            )
+
+            advanceTimeBy(1001)
```

Comment:
For readability and maintenance purposes, I suggest reusing the duration:

```kotlin
val timeUntilExpiration = 1.seconds

// Given
// ...
message = TestMessage.TEXT_MESSAGE.copy(
    expirationData = Message.ExpirationData(
    expireAfter = 1.seconds,
    selfDeletionStatus = Message.ExpirationData.SelfDeletionStatus.NotStarted
)

// end given

// Create an extension function we can reuse for testing
advanceTimeBy(timeUntilExpiration + 1.milliseconds)
```

The extension:
```kotlin
fun TestScope.advanceTimeBy(duration: Duration) = advanceTimeBy(duration.inWholeMilliseconds)
```